### PR TITLE
Update introductory vignette

### PR DESCRIPTION
**Pull request by IndrajeetPatil**
- adds clarifications about inheritance and cloning
- ~~remove hard-coded chunk outputs using `error=TRUE` option in _knitr_~~
- consistently uses `$foo()` convention to refer to class methods
- grammatical changes

### CODE REVIEW

File: vignettes/Introduction.Rmd [hunk 1]
```diff
@@ -180,13 +178,10 @@ n$x
 
 If the function takes no arguments, it's not possible to use it with `<-`:
 
-```{r eval=FALSE}
+```{r error=TRUE}
```
Comment (wch) [inline vignettes/Introduction.Rmd:181]: The reason why it has `eval=FALSE` and the results are hard-coded into the text is because otherwise, the document builds with a different result each time. So it should stay the way it was.

Comment (IndrajeetPatil) [reply vignettes/Introduction.Rmd:181]: I only see one code chunk which makes the document build stochastic. 

````
```{r eval=FALSE}
n$rand
n$rand
n$rand <- 3
```
````

But this can be made reproducible by setting a seed:

````
```{r error=TRUE}
set.seed(123) # for reproducibility

n$rand
n$rand
n$rand <- 3
```
````

Is that an acceptable solution for you?

For the rest of the chunks that had `eval=FALSE`, the output is always the same.

Comment (wch) [reply vignettes/Introduction.Rmd:181]: I don't think adding the `set.seed()` is an improvement over the original version, because it adds more noise for the reader. I suppose you could put it in a block just above with `echo=FALSE`.

File: vignettes/Introduction.Rmd [hunk 2]
```diff
@@ -120,11 +120,9 @@ q$remove()
 
 However, private members can't be accessed directly:
 
-```{r eval = FALSE}
+```{r, error=TRUE}
```
Comment (wch) [inline vignettes/Introduction.Rmd:123]: By evaluating this code chunk, the output here for `q$length()` would change from:

```R
q$length()
#> Error: attempt to apply non-function
```
to :

```R
q$length()
#> Error in eval(expr, envir, enclos): attempt to apply non-function
```

Which I think is a step backward for clarity.

Comment (IndrajeetPatil) [reply vignettes/Introduction.Rmd]: I have reverted changes to all relevant code blocks.

> I think is a step backward for clarity

Sure, the error message might be clearer when hard-coded, but they might also be confusing for some who might wonder why they don't get the exact same error as described in the vignette.

Additionally, if the underlying error description changes in base-R itself, there is no way for the vignette to catch this because the error description is hard-coded.

This is why I had made those changes originally.

Comment (wch) [reply vignettes/Introduction.Rmd]: Just to be clear, the hard-coded error message _is_ the same message that R gives at the console.

image

The message with `Error in eval...` is happens when knitr processes the code chunks.

I suppose it's possible that base R could change the error message at some point in the future, but it seems very unlikely to me.

**Comment by wch** [comment]
Thanks - I think most of the changes are improvements, but there are some places where the change in wording seems more a matter of taste -- they're changes from my writing style to your writing style. I'm not sure what to do about those instances. In most cases I don't feel strongly about it, but there are some cases where I prefer the original wording.

**Comment by IndrajeetPatil** [comment]
Ha, I wish I had a style! 😅

A number of these changes are inspired by grammatical and punctuation suggestions from professional version of [Grammarly](https://www.grammarly.com) software. 

If you feel strongly about certain changes, I am happy to revert them.

File: vignettes/Introduction.Rmd (hunk 2 above)
Comment (IndrajeetPatil) [reply vignettes/Introduction.Rmd]: Okay, fair enough! 

I have already reverted all changes to code chunks with `eval=FALSE`.

**Comment by IndrajeetPatil** [comment]
*bump*

File: vignettes/Introduction.Rmd [hunk 3]
```diff
@@ -11,7 +11,7 @@ The R6 package provides an implementation of encapsulated object-oriented progra
 
 ## R6 classes
 
-R6 classes are similar to R's reference classes, but are lighter weight, and avoid some issues that come along with using S4 (R's reference classes are based on S4). For more information about speed and memory footprint, see `vignette("Performance")`.
+R6 classes are similar to R's reference classes but are smaller in size, and avoid some issues that come along with using S4 (R's reference classes are based on S4). For more information about speed and memory footprint, see `vignette("Performance")`.
```
Comment (wch) [inline vignettes/Introduction.Rmd:14]: Please revert

File: vignettes/Introduction.Rmd [hunk 4]
```diff
@@ -403,9 +423,9 @@ If you don't want a `clone` method to be added, you can use `cloneable=FALSE` wh
 
 #### Deep cloning
 
-If there are any fields which are objects with reference semantics (environments, R6 objects, reference class objects), the copy will get a reference to the same object. This is sometimes desirable, but often it is not.
+If there are any fields that are objects with reference semantics (environments, R6 objects, reference class objects), the copy will get a reference to the same object. This is sometimes desirable, but often it is not.
 
-For example, we'll create an object `c1` which contains another R6 object, `s`, and then clone it. Because the original's and the clone's `s` fields both refer to the same object, modifying it from one results in a change that is reflect in the other.
+For example, we'll create an object `c1` which contains another R6 object, `s`, and then clone it. Because both the original and the clone object's `s` field refers to the same object, modifying it from one results in a change that is reflected in the other.
```
Comment (wch) [inline vignettes/Introduction.Rmd:428]: Please revert, but keep the "reflected" change

File: vignettes/Introduction.Rmd [hunk 5]
```diff
@@ -439,7 +459,7 @@ c1$s$x <- 3
 c3$s$x
 ```
 
-The default behavior of `clone(deep=TRUE)` is to copy fields which are R6 objects, but not copy fields which are environments, reference class objects, or other data structures which contain other reference-type objects (for example, a list with an R6 object).
+The default behavior of `$clone(deep=TRUE)` is to copy fields that are R6 objects, but not to copy fields that are environments, reference class objects, or other data structures which contain other reference-type objects (for example, a list with an R6 object).
```
Comment (wch) [inline vignettes/Introduction.Rmd:462]: Please revert

File: vignettes/Introduction.Rmd [hunk 6]
```diff
@@ -496,11 +516,67 @@ c2$v
 
 In the example `deep_clone` method above, we checked the name of each field to determine what to do with it, but we could also check the value, by using `inherits(value, "R6")`, or `is.environment()`, and so on.
 
+#### Cloning inheritance
+
+If a given R6 class is a subclass, its cloning behavior will be contingent on the cloning behavior of its superclass. Specifically, there are two possibilities:
+
+1. If the superclass has `cloneable = TRUE`, no matter what subclass has set for `cloneable`, cloning will always work.
```
Comment (wch) [inline vignettes/Introduction.Rmd:523]:
```suggestion
1. If the superclass has `cloneable=TRUE`, no matter what subclass has set for `cloneable`, cloning will always work.
```

File: vignettes/Introduction.Rmd [hunk 7]
```diff
@@ -496,11 +516,67 @@ c2$v
 
 In the example `deep_clone` method above, we checked the name of each field to determine what to do with it, but we could also check the value, by using `inherits(value, "R6")`, or `is.environment()`, and so on.
 
+#### Cloning inheritance
```
Comment (wch) [inline vignettes/Introduction.Rmd:519]: I'm not sure that this should be documented at this point -- it is how R6 currently behaves, but that actually feels like a bug to me. At any rate, if we do want to keep this behavior, it should be a separate PR.

File: vignettes/Introduction.Rmd [hunk 8]
```diff
@@ -326,7 +346,7 @@ The new members will be present only in instances that are created after `$set()
 
 To prevent modification of a class, you can use `lock_class=TRUE` when creating the class. You can also lock and unlock a class as follows:
 
-```{r}
+```{r, error=TRUE}
```
[the remark this answers is not in the document]
Comment (IndrajeetPatil) [reply vignettes/Introduction.Rmd:349]: Wouldn't it be better to display the error instead of commenting it out?

Currently, the output on the website:

image

Changed output in the PR:

image

File: vignettes/Introduction.Rmd (hunk 7 above)
Comment (IndrajeetPatil) [reply vignettes/Introduction.Rmd]: I have removed this subsection. Will make a different PR for that change.

**Comment by IndrajeetPatil** [comment]
This should be ready for a merge. I think I've addressed all of your comments.

**Review by wch — APPROVED** [review]
